tests/actor/test_run_utils.py

```diff
@@ -0,0 +1,112 @@
+# License: MIT
+# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
+
+"""Simple tests for the actor runner."""
+
+import time
+from typing import Iterator
+from unittest.mock import patch
+
+import async_solipsism
+import pytest
+import time_machine
+from async_solipsism.socket import asyncio
+
+from frequenz.sdk.actor import actor, run
+
+
+@pytest.fixture(autouse=True)
+def fake_loop() -> Iterator[async_solipsism.EventLoop]:
+    """Replace the loop with one that doesn't interact with the outside world."""
+    loop = async_solipsism.EventLoop()
+    yield loop
+    loop.close()
+
+
+@pytest.fixture
+def fake_time() -> Iterator[time_machine.Coordinates]:
+    """Replace real time with a time machine that doesn't automatically tick."""
+    with time_machine.travel(0, tick=False) as traveller:
+        yield traveller
+
+
+@actor
+class FaultyActor:
+    """A test faulty actor."""
+
+    def __init__(self, name: str) -> None:
+        """Initialize the test faulty actor.
+
+        Args:
+            name: the name of the test faulty actor.
+        """
+        self.name = name
+        self.is_cancelled = False
+
+    async def run(self) -> None:
+        """Run the test faulty actor.
+
+        Raises:
+            CancelledError: after sleeping the duration in seconds set.
+        """
+        self.is_cancelled = True
+        raise asyncio.CancelledError(f"Faulty Actor {self.name} failed")
+
+
+@actor
+class SleepyActor:
+    """A test actor that sleeps a short time."""
+
+    def __init__(self, name: str, duration_sec: float) -> None:
+        """Initialize the test sleepy actor.
+
+        Args:
+            name: the name of the test sleepy actor.
+            duration_sec: the virtual duration in seconds to run.
+        """
+        self.name = name
+        self.duration_sec = duration_sec
+        self.is_joined = False
+        self.virtual_time_taken_to_run = 0.0
+
+    async def run(self) -> None:
+        """Run the test sleepy actor."""
+        start_time = time.time()
+        await asyncio.sleep(self.duration_sec)
+        self.is_joined = True
+        end_time = time.time()
+        self.virtual_time_taken_to_run = end_time - start_time
+        # pylint: disable=protected-access,no-member
+        await self._stop()  # type: ignore[attr-defined]
```

Comment:
Mmm, calling `_stop()` inside `run()` looks fishy. Can't you call it outside of it?

Reply:
yes, you're right. This shouldn't be like this. Either I'll call it from outside or will fix the actor decorator so that an actor is not restarted once it is finished without raising any exception

Reply:
I've added a separate commit to fix the actor decorator, but the changes could be a bit controversial so I'll be also happy to create new issue/PR to discuss about it, and i can call _stop() in the test itself rather than in `run()`